framework/Optimizers/stepManipulators/ConjugateGradient.py

```diff
@@ -0,0 +1,436 @@
+      @ In, recommend, str, optional, override to 'grow' or 'shrink' step size
+      @ In, kwargs, dict, keyword-based specifics as required by individual step sizers
+      @ Out, newOpt, dict, new opt point
+      @ Out, stepSize, float, new step size
+    """
+    # Conjugate Gradient does line searches along consecutive gradient estimations
+    ## with gradient estimations updated by more than just local estimation.
+    ## For conjugate gradient notations, see https://en.wikipedia.org/wiki/iNonlinear_conjugate_gradient_method
+    ## For line search notations, see github.com/scipy/scipy/blob/master/scipy/optimize/minpack2/dcsrch.f
+
+    # We start from an opt point, then find the gradient direction.
+    #   from there we line search for the best point along that grad direction
+    #   During this time, we store the original opt point from were we found the gradient,
+    #     as well as the best point found so far along the line search. -> optPointHist
+    # In the grad hist, we store only gradients around best-in-line points historically
+    # In the step hist, we store line search information
+
+    lastStepInfo = prevStepSize[-1]['info']
+    if lastStepInfo is None:
+      # this MUST MEAN that this is the very very first step in this algorithm
+      # note that we use binary strings because that's what scipy gives us
+      lastStepInfo = {'task': b'START',
+                      'fortranParams': {'iSave': np.zeros((2,), dtype=np.intc),
+                                        'dSave': np.zeros((13,), dtype=float),
+                                       }
+                     }
+    lineSearchTask = lastStepInfo['task']
+
+    # get the gradient
+    curGradMag = gradientHist[-1][0]
+    curGrad = np.array(list(gradientHist[-1][1][v] for v in self._optVars)) * curGradMag
+
+    # get an array of the current optimal point
+    curPoint = np.array(list(prevOpt[v] for v in self._optVars))
+    # get the current opt point objective value
+    curObjVal = prevOpt[objVar]
+
+    # if we're starting a new line search because we found a minimum along the previous line search
+    ## NOTE this only gets called the first time ever for each trajectory, because of how we start
+    ##      new line searches under the task == 'CONVERGE' switch below
+    if lineSearchTask == b'START':
+      lastStepInfo = self._startLineSearch(lastStepInfo, curPoint, curObjVal, curGrad, curGradMag)
+    else: # if lineSearchTask is anything except "start"
+      # store some indicative information about the gradient (scalar product of current gradient and
+      # the search vector, also "derPhi" in literature)
+      lastStepInfo['line']['objDerivative'] = np.dot(curGrad, lastStepInfo['searchVector']) # derPhi1
+
+    # update the line search information, and get the next task (and step size if relevant)
+    stepSize, task = self._lineSearchStep(lastStepInfo, curObjVal)
+
+    # take actions depending on the task
+    if task.startswith(b'FG'):
```

Comment:
where are "these tasks" explained?

Reply:
scipy source code (fortran), which is noted in the comments in this method. See line 133, where I even give the website to go to with all the notations, in case they change in future versions of scipy.